kernel/src/table_changes/mod.rs

```diff
@@ -90,7 +90,7 @@ impl TableChanges {
         // supported for every protocol action in the CDF range.
         let start_snapshot =
             Snapshot::try_new(table_root.as_url().clone(), engine, Some(start_version))?;
-        let end_snapshot = Snapshot::try_new(table_root.as_url().clone(), engine, end_version)?;
+        let end_snapshot = Snapshot::new_from(&start_snapshot, engine, end_version)?;
```

Comment:
This opens an interesting question... if we _knew_ that `new_from` would reuse the log checkpoint and just "append" any new commit .json files to the log segment, then we could almost (**) reuse that log segment for the CDF replay by just stripping out its checkpoint files? But that's pretty CDF specific; in the normal case we want a refresh to use the newest checkpoint available because it makes data skipping log replay cheaper. Maybe the CDF case needs a completely different way of creating the `end_snapshot`, unrelated to this optimization here.

(**) Almost, because the start version might have a checkpoint, in which case stripping the checkpoint out of the log segment would also remove the start version. But then again, do we _actually_ want the older snapshot to be the start version? Or the previous version which the start version is making changes to? Or, maybe we should just restrict the checkpoint search to versions before the start version, specifically so that this optimization can work.

Reply:
> do we actually want the older snapshot to be the start version?

It would be sufficient to have the older snapshot be `start_version-1` as long as we also have access to the commit at `start_version`. With these, we would start P&M at `start_version` then continue it on the older snapshot if we don't find anything.

I guess this would look like: `snapshot(start_version-1).refresh_with_commits(end_version)`

After all, the goal of the start_snapshot is just to ensure that CDF is enabled.